src/main/scala/rules/Joiner.scala

```diff
@@ -13,18 +13,24 @@ import viper.silicon.logger.records.structural.JoiningRecord
 import viper.silicon.state.State
 import viper.silicon.verifier.Verifier
 
-case class JoinDataEntry[D](s: State, data: D, pathConditions: RecordedPathConditions)
+case class JoinDataEntry[D](s: State, data: D, pathConditions: RecordedPathConditions) {
+  // Instead of merging states, we can directly merge JoinDataEntries to obtain new States.
```

Comment:
Is this a "future work" comment? Otherwise, I'm confused: you say "instead of merging states", but then have `State.merge` on line 20.